File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        //two pointer approach
        int left = 0;
        int right = height.size() - 1;
        int maxRight = 0;
        int maxLeft = 0;
        int water = 0;

        while (left < right) {
            if (height[left] <= height[right]) {
                //now we consider the left max
                maxLeft = max(height[left], maxLeft);
                water += maxLeft - height[left]; //minimum is 0
                left++;
            }
            else {
                maxRight = max(height[right], maxRight);
                water += maxRight - height[right];
                right--;
            }
        }
        return water;
        
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        right = n * [1]
        left = n * [1]
        
        #right array
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]
        
        #left array
        for i in range(1, n):
            left[i] = left[i - 1] * nums[i - 1]
            print(left[i])

        result = [right[i] * left[i] for i in range(len(nums))]
        return result
        
       


        